src/Storages/FileLog/StorageFileLog.cpp

```diff
@@ -0,0 +1,445 @@
+
+    return std::min(batch_size, getMaxBlockSize());
+}
+
+size_t StorageFileLog::getPollTimeoutMillisecond() const
+{
+    return filelog_settings->filelog_poll_timeout_ms.changed ? filelog_settings->filelog_poll_timeout_ms.totalMilliseconds()
+                                                             : getContext()->getSettingsRef().stream_poll_timeout_ms.totalMilliseconds();
+}
+
+
+bool StorageFileLog::checkDependencies(const StorageID & table_id)
+{
+    // Check if all dependencies are attached
+    auto dependencies = DatabaseCatalog::instance().getDependencies(table_id);
+    if (dependencies.empty())
+        return true;
+
+    for (const auto & storage : dependencies)
+    {
+        auto table = DatabaseCatalog::instance().tryGetTable(storage, getContext());
+        if (!table)
+            return false;
+
+        // If it materialized view, check it's target table
+        auto * materialized_view = dynamic_cast<StorageMaterializedView *>(table.get());
+        if (materialized_view && !materialized_view->tryGetTargetTable())
+            return false;
+
+        // Check all its dependencies
+        if (!checkDependencies(storage))
+            return false;
+    }
+
+    return true;
+}
+
+void StorageFileLog::threadFunc()
+{
+    try
+    {
+        auto table_id = getStorageID();
+        // Check if at least one direct dependency is attached
+        size_t dependencies_count = DatabaseCatalog::instance().getDependencies(table_id).size();
+        if (dependencies_count)
+        {
+            auto start_time = std::chrono::steady_clock::now();
+
+            watch_task->activateAndSchedule();
+
+            // Keep streaming as long as there are attached views and streaming is not cancelled
+            while (!task->stream_cancelled)
+            {
+                if (!checkDependencies(table_id))
+                    break;
+
+                LOG_DEBUG(log, "Started streaming to {} attached views", dependencies_count);
+
+                auto stream_is_stalled = streamToViews();
+                if (stream_is_stalled)
+                {
+                    LOG_TRACE(log, "Stream stalled. Reschedule.");
+                    break;
+                }
+
+                auto ts = std::chrono::steady_clock::now();
+                auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(ts-start_time);
+                if (duration.count() > MAX_THREAD_WORK_DURATION_MS)
+                {
+                    LOG_TRACE(log, "Thread work duration limit exceeded. Reschedule.");
+                    break;
+                }
+            }
+        }
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+
+    watch_task->deactivate();
+    // Wait for attached views
+    if (!task->stream_cancelled)
+        task->holder->scheduleAfter(RESCHEDULE_MS);
+}
+
+
+bool StorageFileLog::streamToViews()
+{
+    std::lock_guard<std::mutex> lock(status_mutex);
```

Comment:
Having this mutex for the whole duration of streamToViews does not seem correct. As I see we will not be able to read anything because we will wait under the same mutex in watchFunc.

Reply:
But the mutex not always hold by watchFunc?

Reply:
Yes, but files statuses are not updated for the whole duration of streamToViews method. Then such scheme does not seem efficient and it will be better if we update statuses just after each streamToViews iteration -- sequentially.
Also I had an idea that we can try to allow updating file statuses while streamToViews iteration is in execution. My idea was to have global file statuses and local file statuses. Local files statuses are statuses that each FileLogSource currently has. Each FileLogSource steps into cycle in generate() and does not update its local file statuses untill next call to generate. At the beginning of generate() FileLogSource checks under mutex global file statuses and updates his own local file statuses.
What do you think?

Reply:
For the second method with local files statuses, how about new added files during the streamToViews method?

Reply:
I think update statuses after each streamToViews iteration is better, BTW, we can use have files updated/added to indicate that the stream does not stalled, if no update, the task should be reschedule?

Reply:
I think update statuses after each streamToViews iteration is better, but if no mv attached, the file statuses will never be updated.  BTW, we can use have files updated/added to indicate that the stream does not stalled, if no update, the task should be reschedule?

Reply:
> I think update statuses after each streamToViews iteration is better

Yes, I also think it will be ok.

> BTW, we can use have files updated/added to indicate that the stream does not stalled, if no update, the task should be reschedule?

Yes

Reply:
We can remove this mutex now as I see.